src/ILCompiler.ReadyToRun/src/ObjectWriter/Relocator.cs

```diff
@@ -0,0 +1,559 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+
+namespace ILCompiler.PEWriter
+{
+    public static class RelocType
+    {
+        /// <summary>
+        /// No relocation required
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ABSOLUTE = 0x00;
+
+        /// <summary>
+        /// The 32-bit address without an image base (RVA)
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ADDR32NB = 0x02;
+
+        /// <summary>
+        /// 32 bit address base
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_HIGHLOW = 0x03;
+
+        /// <summary>
+        /// Thumb2: based MOVW/MOVT
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_THUMB_MOV32 = 0x07;
+
+        /// <summary>
+        /// 64 bit address base
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_DIR64 = 0x0A;
+
+        /// <summary>
+        /// 32-bit relative address from byte following reloc
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_REL32 = 0x10;
+
+        /// <summary>
+        /// Thumb2: based B, BL
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_THUMB_BRANCH24 = 0x13;
+
+        /// <summary>
+        /// Arm64: B, BL
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_BRANCH26 = 0x14;
+
+        /// <summary>
+        /// 32-bit relative address from byte starting reloc
+        /// This is a special NGEN-specific relocation type
+        /// for relative pointer (used to make NGen relocation
+        /// section smaller)
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_RELPTR32 = 0x7C;
+
+        /// <summary>
+        /// 32 bit offset from base of section containing target
+        /// </summary>
+        public const ushort IMAGE_REL_SECREL = 0x80;
+
+        /// <summary>
+        /// ADRP
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEBASE_REL21 = 0x81;
+
+        /// <summary>
+        /// ADD/ADDS (immediate) with zero shift, for page offset
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEOFFSET_12A = 0x82;
+
+        /// <summary>
+        /// LDR (indexed, unsigned immediate), for page offset
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEOFFSET_12L = 0x83;
+    }
+
+    /// <summary>
+    /// Abstract class representing a relocator i.e. an algorithm capable of
+    /// modifying a sequence of bytes by applying the given relocation type
+    /// with RVA and delta parameters.
+    /// </summary>
+    public abstract class Relocator
```

Comment:
I think having abstract class to represent Relocation algorithm is overkill. Enum should work just fine.

Reply:
Removed and replaced with a simple switch in 2nd commit.